nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,567 @@
+    in nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name}";
+      path = with pkgs; [
+        borgbackup openssh
+      ];
+      script = mkBackupScript cfg;
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Only run when no other process is using CPU or disk
+        CPUSchedulingPolicy = "idle";
+        IOSchedulingClass = "idle";
+        ProtectSystem = "strict";
+        ReadWritePaths =
+          [ "${userHome}/.config/borg" "${userHome}/.cache/borg" ]
+          # Borg needs write access to repo if it is not remote
+          ++ optional (isPath cfg.repo) cfg.repo;
+        PrivateTmp = true;
+      };
+      environment = {
+        BORG_REPO = cfg.repo;
+        inherit (cfg) archiveName;
+        inherit (cfg) extraInitArgs extraCreateArgs extraPruneArgs;
+      } // (mkPassEnv cfg) // cfg.environment;
+      inherit (cfg) startAt;
+    };
+
+  # Paths listed in ReadWritePaths must exist before service is started
+  mkActivationScript = name: cfg:
+    let
+      install = "install -o ${cfg.user} -g ${cfg.group}";
+    in
+      nameValuePair "borgbackup-job-${name}" (stringAfter [ "users" ] (
+        ''
+          # Eensure that the home directory already exists
+          # We can't assert createHome == true because that's not the case for root
+          cd "${config.users.users.${cfg.user}.home}"                                                                                                         
+          ${install} -d .config/borg
+          ${install} -d .cache/borg
+        '' + optionalString (isPath cfg.repo) ''                                                                                                              
+          ${install} -d ${escapeShellArg cfg.repo}
+        ''
+      ));
+
+  mkPassAssertion = name: cfg: {
+    assertion = with cfg.encryption;
+      mode != "none" -> passCommand != null || passphrase != null;
+    message =
+      ''passCommand or passphrase have to be specified ''
+      + ''because borgbackup.jobs.${name}.encryption != "none"'';
+  };
+
+  mkRepoService = name: cfg:
+    nameValuePair "borgbackup-repo-${name}" {
+      description = "Create BorgBackup repository ${name} directory";
+      script = ''
+        mkdir -p ${escapeShellArg cfg.path}
+        chown ${cfg.user}:${cfg.group} ${escapeShellArg cfg.path}
+      '';
+      serviceConfig = {
+        # The service's only task is to ensure that the specified path exists
+        Type = "oneshot";
+      };
+      wantedBy = [ "multi-user.target" ];
+    };
+
+  mkAuthorizedKey = cfg: appendOnly: key:
+    let
+      restrictedArg = "--restrict-to-repository ${escapeShellArg cfg.path}";
+      appendOnlyArg = optionalString appendOnly "--append-only";
+      quotaArg = optionalString (cfg.quota != null) "--storage-quota ${cfg.quota}";
+      command = "borg serve ${restrictedArg} ${appendOnlyArg} ${quotaArg}";
```

Comment:
Feature request. Add optional "cd ${cfg.path}" in front of that "borg server" (Inspired by http://borgbackup.readthedocs.io/en/stable/deployment/central-backup-server.html#restrictions.) This allows clients to ignore the server directory layout, and can use repo urls like backup@server:myrepo.

Reply:
Nice idea!
Btw, the example you linked to uses `--restrict-to-path` instead of `--restrict-to-repository`. So, as it is currently implemented there is no way of using multiple repos with the same key (except if you use multiple users on the storage server).

Reply:
Do you have a use case for `--restrict-to-path`?
I implement the working directory change, but noticed the way you specify the repo is quite ugly (`user@machine:.`).

Reply:
Re: --restrict-to-path. I use that in my local setup (not yet pushed to github). I chose that option because it allowed the client to create the repository it needs. I have something like /data/borgbackup/hosts/machine1 and .../machine2, where each machine may have a number of repositories, managed by the client. I'm currently experimenting with the setup, hence it's not pushed yet. I don't have a good answer. I guess I can flip it around and ask "what's the use case of --restrict-to-repo"?

Ah, it should be "cd ${config.services.borgbackup.repos.${name}.path}". (Or whatever expands to the path where repos are stored, e.g. /var/lib/borgbackup.) Then we can specify repo correctly, I think, without the "." (dot).

Btw, I think it's a good idea to "cd dir && borg serve ...", so that borg isn't run if the cd fails. (Unless that's considered a feature in this setup -- I didn't check.)

Reply:
You're right, there's not much use in `--restrict-to-repository` for restricting repository access. However, I think `--restrict-to-path` makes `--storage-quota` useless.

No, we can't leave out the dot:
```
$ BORG_REPO="machine:" borg create ::archive path
Repository /home/user/machine: does not exist.
```

Reply:
I think the best solution is an option called `allowSubRepos`. Do you think it should be `true` or `false` by default?

Reply:
Good point about the --storage-quota. Ok, now I think I understand why there are two --restrict-to-* options :-)

Re: the dot. I was thinking of this syntax: ```BORG_REPO="machine:repo.borg" borg create ::archive path```.

> I think the best solution is an option called allowSubRepos. Do you think it should be true or false by default?

No opinion yet. Do you plan to select the --restrict-to-* option based on allowSubRepos?

Reply:
Yes exactly. See the third commit.

Reply:
Sweet.

Reply:
Dude.